### ISSUE TITLE
Rule S3903: False Positive when using top level statements and partial Program in separate file

### ISSUE THREAD

**Issue by dnperfors** [SonarSource/sonar-dotnet#6836]
### Description

This issue is similar to #5660, which was fixed some time ago.

Because we have several attributes on the Program class (to suppress messages, to exclude code coverage) and we needed to make the class public, we opted to place this partial class in a separate file. Today I noticed we still hade the suppression that was solved in #5660, so I though "Let's remove that, since it is fixed and our SonarQube is updated". To my surprise the warning came back.
After fiddling around, I noticed that this only happened when the partial class is in a separate file. When the partial class is in the Program.cs after the top level statements the warning is not there.

As described in the previous issue, Roslyn has a similar warning ([CA1050](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1050)), but that one doesn't give a warning when the partial class is in a separate file.

### Repro steps

In Program.cs:
```
Console.Write("Hello World");
```
In Program.Attributes.cs:
```
[ExcludeFromCodeCoverage]
public partial class Program
{
    private Program() { }
}
```

Warning given:
```
##[warning]Program.Attributes.cs(2,22): Warning S3903: Move 'Program' into a named namespace.

### Expected behavior

The [S3903](https://sonarsource.github.io/rspec/#/rspec/S3903/csharp) warning is not emitted.

### Actual behavior

The [S3903](https://sonarsource.github.io/rspec/#/rspec/S3903/csharp) warning is emitted

### Known workarounds

Place the partial class in the same file as described in #5660

### Related information

* C# Plugin version: 8.46
* MSBuild / dotnet version: dotnet 7.0

**Comment by mary-georgiou-sonarsource** [comment]
Thank you for opening this issue, @dnperfors.
I can confirm that this is indeed a False Positive; I added it to our backlog to be tackled in a future sprint.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1840](https://sonarsource.atlassian.net/browse/NET-1840)

[NET-1840]: https://sonarsource.atlassian.net/browse/NET-1840?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
